contracts/src/v0.8/keystone/KeystoneFeedsConsumer.sol

```diff
@@ -2,27 +2,92 @@
+  bytes10[] internal s_allowedWorkflowNamesList;
+  mapping(bytes10 => bool) internal s_allowedWorkflowNames;
+
+  function setConfig(address[] calldata _allowedSendersList, address[] calldata _allowedWorkflowOwnersList, bytes10[] calldata _allowedWorkflowNamesList) external onlyOwner {
+    for (uint32 i = 0; i < s_allowedSendersList.length; i++) {
+      s_allowedSenders[s_allowedSendersList[i]] = false;
+    }
+    for (uint32 i = 0; i < _allowedSendersList.length; i++) {
+      s_allowedSenders[_allowedSendersList[i]] = true;
+    }
+    s_allowedSendersList = _allowedSendersList;
+    for (uint32 i = 0; i < s_allowedWorkflowOwnersList.length; i++) {
+      s_allowedWorkflowOwners[s_allowedWorkflowOwnersList[i]] = false;
+    }
+    for (uint32 i = 0; i < _allowedWorkflowOwnersList.length; i++) {
+      s_allowedWorkflowOwners[_allowedWorkflowOwnersList[i]] = true;
+    }
+    s_allowedWorkflowOwnersList = _allowedWorkflowOwnersList;
+    for (uint32 i = 0; i < s_allowedWorkflowNamesList.length; i++) {
+      s_allowedWorkflowNames[s_allowedWorkflowNamesList[i]] = false;
+    }
+    for (uint32 i = 0; i < _allowedWorkflowNamesList.length; i++) {
+      s_allowedWorkflowNames[_allowedWorkflowNamesList[i]] = true;
+    }
+    s_allowedWorkflowNamesList = _allowedWorkflowNamesList;
+  }
+
+  function onReport(bytes calldata metadata, bytes calldata rawReport) external {
+    if (s_allowedSenders[msg.sender] == false) {
+      revert UnauthorizedSender(msg.sender);
+    }
+
+    (bytes10 workflowName, address workflowOwner) = _getInfo(metadata);
+    if (s_allowedWorkflowNames[workflowName] == false) {
+      revert UnauthorizedWorkflowName(workflowName);
+    }
+    if (s_allowedWorkflowOwners[workflowOwner] == false) {
+      revert UnauthorizedWorkflowOwner(workflowOwner);
+    }
 
     FeedReport[] memory feeds = abi.decode(rawReport, (FeedReport[]));
     for (uint32 i = 0; i < feeds.length; i++) {
-      emit FeedReceived(feeds[i].FeedID, feeds[i].Price, feeds[i].Timestamp);
+      s_prices[feeds[i].FeedId] = feeds[i].Price;
+      emit FeedReceived(feeds[i].FeedId, feeds[i].Price, feeds[i].Timestamp);
     }
+  }
+
+  function _getInfo(
+    bytes memory metadata
+  ) internal pure returns (bytes10 workflowName, address workflowOwner) {
+    // (first 32 bytes contain length of the byte array)
+    // workflow_cid             // offset 32, size 32
+    // workflow_name            // offset 64, size 10
+    // workflow_owner           // offset 74, size 20
+    // report_name              // offset 94, size  2
+    assembly {
+      // shift right by 22 bytes to get the actual value
+      workflowName := shr(mul(22, 8), mload(add(metadata, 64)))
+      // shift right by 12 bytes to get the actual value
+      workflowOwner := shr(mul(12, 8), mload(add(metadata, 74)))
+    }
+  }
 
-    emit MessageReceived(workflowId, workflowOwner, feeds.length);
+  function getPrice(bytes32 feedId) external view returns (uint256) {
```

Comment:
Should probably be updated to include returning the timestamp the price was recorded at so that whomever calls this contract can check for staleness.